README
The BitVector.py module is for a memory-efficient packed
representation of bit vectors and for the processing of such
vectors.

If you wish, you can execute the code in

     BitVector.py

directly in this directory prior to the installation of the
module.  This will execute all of the example code in the
module file.

To see a working examples of the BitVector module, see 
the file

     Examples/BitVectorDemo.py

If there are any questions about this module, contact the
author at 'kak@purdue.edu' with the string 'bitvector' in
the subject line to get past his spam filter.

HOW THE BIT VECTORS ARE STORED:

The bits of a bit vector are stored in 16-bit unsigned ints
following Josiah Carlson's recommendation to that effect on the
Pyrex mailing list.  As you can see in the code for `__init__()',
after resolving the argument with which the constructor is called,
the very first thing the constructor does is to figure out how many
of those 2-byte ints it needs for the bits (see how the value is
assigned to the variable `two_byte_ints_needed' toward the end of
`__init__()').  For example, if you wanted to store a 64-bit array,
the variable 'two_byte_ints_needed' would be set to 4. (This does
not mean that the size of a bit vector must be a multiple of 16.
Any sized bit vectors can be constructed --- the constructor will
choose the minimum number of two-byte ints needed.) Subsequently,
the constructor acquires an array of zero-initialized 2-byte ints.
The last thing that is done in the code for `__init__()' is to
shift the bits into the array of two-byte ints.

As mentioned above, note that it is not necessary for the size of a
bit vector to be a multiple of 16 even though we are using C's
unsigned short as as a basic unit for storing the bit arrays.  The
class BitVector keeps track of the actual number of bits in the bit
vector through the "size" instance variable.

Note that, except for one case, the constructor must be called with
a single keyword argument, which determines how the bit vector will
be constructed.  The single exception to this rule is for the
keyword argument `intVal' which can be used along with the `size'
keyword argument.  When `intVal' is used without the `size' option,
the bit vector constructed for the integer is the shortest possible
bit vector.  On the other hand, when `size' is also specified, the
bit vector is padded with zeroes from the left so that it has the
specified size.  The code for `__init__()' begins by making sure
your constructor call only uses the acceptable keywords.  The
constraints on how many keywords can be used together in a
constructor call are enforced when we process each keyword option
separately in the rest of the code for `__init__()'.

The first keyword option processed by `__init__()' is for
`filename'.  When the constructor is called with the `filename'
keyword, as in

       bv = BitVector(filename = 'myfilename')

the call returns a bit vector on which you must subsequently invoke
the `read_bits_from_file()' method to actually obtain a bit vector
consisting of the bits that constitute the information stored in
the file.

The next keyword option considered in `__init__()' is for `fp',
which is for constructing a bit vector by reading off the bits from
a file-like object, as in

      x = '111100001111'
      fileobj = StringIO.StringIO( x )
      bv = BitVector( fp = fileobj )

The keyword option `intVal' considered next is for converting an
integer into a bit vector through a constructor call like

      bv = BitVector(intVal = 123456)

The bits stored in the bit vector thus created correspond to the
big-endian binary representation of the integer argument provided
through `intVal' (meaning that the most significant bit will be at
the leftmost position in the bit vector.)  THE BIT VECTOR
CONSTRUCTED WITH THE ABOVE CALL IS THE SHORTEST POSSIBLE BIT VECTOR
FOR THE INTEGER SUPPLIED.  As a case in point, when `intVal' is set
to 0, the bit vector consists of a single bit is 0 also.  When
constructing a bit vector with the `intVal' option, if you also
want to impose a size condition on the bit vector, you can make a
call like

      bv = BitVector(intVal = 46, size = 16)

which returns a bit vector of the indicated size by padding the
shortest possible vector for the `intVal' option with zeros from
the left.

The next option processed by `__init_()' is for the `size' keyword
when this keyword is used all by itself.  If you want a bit vector
of just 0's of whatever size, you make a call like

      bv = BitVector(size = 61)

This returns a bit vector that will hold exactly 61 bits, all
initialized to the zero value.

The next constructor keyword processed by `__init__()' is
`bitstring'. This is to allow a bit vector to be constructed
directly from a bit string as in

      bv = BitVector(bitstring = '00110011111')

The keyword considered next is `bitlist' which allows a bit vector
to be constructed from a list or a tuple of individual bits, as in

      bv = BitVector(bitlist = (1, 0, 1, 1, 0, 0, 1))

The last two keyword options considered in `__init__()' are for
keywords `textstring' and `hexstring'.  If you want to construct a
bitvector directly from a text string, you call

      bv = BitVector(textstring = 'hello')

The bit vector created corresponds to the ASCII encodings of the
individual characters in the text string.

And if you want to do the same with a hex string, you call

      bv = BitVector(hexstring = '68656c6c6f')

Now, as you would expect, the bits in the bit vector will
correspond directly to the hex digits in your hex string.

OPERATIONS SUPPORTED BY THE BITVECTOR CLASS:

DISPLAYING BIT VECTORS:

@tag1
(1) Since the BitVector class implements the __str__ method, a bit
    vector can be displayed on a terminal by

          print(bitvec)

    or, for only Python 2.x, by

          print bitvec

    Basically, you can always obtain the string representation of a
    bit vector by

          str(bitvec)

    and integer value by

          int(bitvec)


ACCESSING AND SETTING INDIVIDUAL BITS AND SLICES:

@tag2
(2) Any single bit of a bit vector bv can be set to 1 or 0 by

           bv[M] = 1_or_0
           print(bv[M])

     or, for just Python 2.x, by

           bv[M] = 1_or_0
           print bv[M]

     for accessing (and setting) the bit at the position that is
     indexed M.  You can retrieve the bit at position M by bv[M].
     Note that the index 0 corresponds to the first bit at the left
     end of a bit pattern.  This is made possible by the
     implementation of the __getitem__ and __setitem__ methods.

 @tag3
 (3) A slice of a bit vector obtained by

           bv[i:j]

     is a bit vector constructed from the bits at index positions
     from i through j-1.  This is made possible by the
     implementation of the __getslice__ method.

 @tag4
 (4) You can also carry out slice assignment:

           bv1 = BitVector(size = 25)
           bv2 = BitVector(bitstring = '1010001')
           bv1[6:9]  = bv2[0:3]
           bv3 = BitVector(bitstring = '101')
           bv1[0:3]  = bv3

     The first slice assignment will set the 6th, 7th, and the 8th
     bits of the bit vector bv1 according to the first three bits of
     bv2.  The second slice assignment will set the first three bits
     of bv1 according to the three bits in bv3.  This is made
     possible by the slice setting code in the __setitem__ method.

 @tag5
 (5) You can iterate over a bit vector, as illustrated by

           for bit in bitvec:
               print(bit)

     This is made possible by the override definition for the special
     __iter__() method.

 @tag6
 (6) Negative subscripts for array-like indexing are supported.
     Therefore,

           bitvec[-i]

     is legal assuming that the index range is not violated.  A
     negative index carries the usual Python interpretation: The
     last element of a bit vector is indexed -1 and the first
     element -(n+1) if n is the total number of bits in the bit
     vector.  Negative subscripts are made possible by
     special-casing such access in the implementation of the
     __getitem__ method (actually it is the _getbit method).

 @tag7
 (7) You can reset a previously constructed bit vector to either the
     all-zeros state or the all-ones state by

           bv1 = BitVector(size = 25)
           ...
           ...
           bv1.reset(1)
           ...
           ...
           bv1.reset(0)

     The first call to reset() will set all the bits of bv1 to 1's
     and the second call all the bits to 0's.


tle
ICAL OPERATIONS ON BIT VECTORS:

 @tag8
 (8) Given two bit vectors bv1 and bv2, you can perform bitwise
     logical operations on them by

           result_bv  =  bv1 ^ bv2           # for bitwise XOR
           result_bv  =  bv1 & bv2           # for bitwise AND
           result_bv  =  bv1 | bv2           # for bitwise OR
           result_bv  =  ~bv1                # for bitwise negation

     These are made possible by implementing the __xor__, __and__,

ChangeLog
Version 3.3.2:

     This version fixes a bug in the constructor code for creating a
     bit vector from a text string.  The bug was triggered by
     character escapes in such strings.

Version 3.3.1:

    This is a minor upgrade to make the syntax of the API method
    declarations more uniform.  Previously, while most of the method
    names used underscores to connect multiple words, some used
    camelcasing.  Now all use underscores.  For backward
    compatibility, the old calls will continue to work.

Version 3.3:

    This version includes: (1) One additional constructor mode that
    allows a bit vector to be constructed directly from the bytes
    type objects in the memory. (2) A bugfix in the slice function
    for the case when the upper and the lower bounds of the slice
    range are identical. (3) A bugfix for the next_set_bit() method.

Version 3.2:

    This version includes support for constructing bit vectors
    directly from text strings and hex strings.  This version also
    includes a safety check on the sizes of the two argument bit
    vectors when calculating Jaccard similarity between the two.

Version 3.1.1:

    This version includes: (1) a fix to the module test code to
    account for how string input is handled in the io.StringIO class
    in Python 2.7; (2) some improvements to the documentation.

Version 3.1:

    This version includes: (1) Correction for a documentation error;
    (2) Fix for a bug in slice assignment when one or both of the
    slice limits were left unspecified; (3) The non-circular bit
    shift methods now return self so that they can be chained; (4) A
    method for testing a bitvector for its primality; and (5) A
    method that uses Python's 'random.getrandbits()' to generate
    a bitvector that can serve as candidate for primes whose bitfield
    size is specified.

Version 3.0:

    This is a Python 3.x compliant version of the latest incarnation
    of the BitVector module.  This version should work with both
    Python 2.x and Python 3.x.

Version 2.2:

    Fixed a couple of bugs, the most important being in the
    bitvector initialization code for the cases when the
    user-specified value for size conflicts with the user-specified
    int value for the vector.  Version 2.2 also includes a new
    method runs() that returns a list of strings of the consecutive
    runs of 1's and 0's in the bitvector.  The implementation of
    the circular shift operators has also been improved in Version
    2.2. This version allows for a chained invocation of these
    operators.  Additionally, the circular shift operators now
    exhibit expected behavior if the user-specified shift value is
    negative.

Version 2.1:

    Includes enhanced support for folks who use this class for
    computer security and cryptography work.  You can now call on
    the methods of the BitVector class to do Galois Field GF(2^n)
    arithmetic on bit arrays.  This should save the users of this
    class the bother of having to write their own routines for
    finding multiplicative inverses in GF(2^n) finite fields.

Version 2.0.1:

    Fixed numerous typos and other errors in the documentation page
    for the module.  The implementation code remains unchanged.

Version 2.0:

    To address the needs of the folks who are using the BitVector
    class in data mining research, the new version of the class
    includes several additional methods.  Since the bitvectors used
    by these folks can be extremely long, possibly involving
    millions of bits, the new version of the class includes a much
    faster method for counting the total number of set bits when a
    bitvector is sparse.  [But note that this new bit counting
    method may perform poorly for dense bitvectors. So the old bit
    counting method has been retained.]  Also for data mining folks,
    the new version of the class is provided with similarity and
    distance calculation metrics such as the Jaccard similarity
    coefficient, the Jaccard distance, and the Hamming distance.
    Again for the same folks, the class now also has a
    next_set_bit(from_index) method.  Other enhancements to the
    class include methods for folks who do research in cryptography.
    Now you can directly calculate the greatest common divisor of
    two bitvectors, or find the multiplicative inverse of one
    bitvector modulo another bitvector.

Version 1.5.1:

    Removed a bug from the implementation of the right circular
    shift operator.


Examples/BitVectorDemo.py
#!/usr/bin/env python

from BitVector import *

# Construct an EMPTY bit vector (a bit vector of size 0):
print("\nConstructing an EMPTY bit vector (a bit vector of size 0):")
bv1 = BitVector( size = 0 )
print(bv1)                                   # no output

# Construct a bit vector of size 2:
print("\nConstructing a bit vector of size 2:")
bv2 = BitVector( size = 2 )
print(bv2)                                   # 00

# Joining two bit vectors:
print("\nOutput concatenation of two previous bit vectors:")
result = bv1 + bv2
print(result)                                # 00

# Construct a bit vector with a tuple of bits:
print("\nThis is a bit vector from a tuple of bits:")
bv = BitVector(bitlist=(1, 0, 0, 1))
print(bv)                                    # 1001

# Construct a bit vector with a list of bits:
print("\nThis is a bit vector from a list of bits:")
bv = BitVector(bitlist=[1, 1, 0, 1])
print(bv)                                    # 1101

# Construct a bit vector from an integer
bv = BitVector(intVal=5678)
print("\nBit vector constructed from integer 5678:")
print(bv)                                    # 1011000101110
print("\nBit vector constructed from integer 0:")
bv = BitVector(intVal=0)
print(bv)                                    # 0
print("\nBit vector constructed from integer 2:")
bv = BitVector(intVal=2)
print(bv)                                    # 10
print("\nBit vector constructed from integer 3:")
bv = BitVector(intVal=3)
print(bv)                                    # 11
print("\nBit vector constructed from integer 123456:")
bv = BitVector(intVal=123456)
print(bv)                                    # 11110001001000000
print("\nInt value of the previous bit vector as computed by intVal():")
print(bv.intValue())                         # 123456
print("\nInt value of the previous bit vector as computed by int():")
print(int(bv))                               # 123456

# Construct a bit vector from a very large integer:
x = 12345678901234567890123456789012345678901234567890123456789012345678901234567890
bv = BitVector(intVal=x)
print("\nHere is a bit vector constructed from a very large integer:")
print(bv)
print("The integer value of the above bit vector is:%d" % int(bv))

# Construct a bit vector directly from a file-like object:
import io
x = "111100001111"
x = ""
if sys.version_info[0] == 3:
    x = "111100001111"
else:
    x = unicode("111100001111")
fp_read = io.StringIO(x)
bv = BitVector( fp = fp_read )
print("\nBit vector constructed directed from a file like object:")
print(bv)                                    # 111100001111

# Construct a bit vector directly from a bit string:
bv = BitVector( bitstring = '00110011' )
print("\nBit Vector constructed directly from a bit string:")
print(bv)                                    # 00110011

bv = BitVector(bitstring = '')
print("\nBit Vector constructed directly from an empty bit string:")
print(bv)                                    # nothing
print("\nInteger value of the previous bit vector:")
print(bv.intValue())                         # 0

print("\nConstructing a bit vector from the textstring 'hello':")
bv3 = BitVector(textstring = "hello")
print(bv3)
mytext = bv3.getTextFromBitVector()
print("Text recovered from the previous bitvector: ")
print(mytext)                                         # hello
print("\nConstructing a bit vector from the textstring 'hello\\njello':")
bv3 = BitVector(textstring = "hello\njello")
print(bv3)
mytext = bv3.get_text_from_bitvector()
print("Text recovered from the previous bitvector:")
print(mytext)                                         # hello
                                                      # jello
print("\nConstructing a bit vector from the hexstring '68656c6c6f':")
bv4 = BitVector(hexstring = "68656c6c6f")
print(bv4)
myhexstring = bv4.getHexStringFromBitVector()
print("Hex string recovered from the previous bitvector: ")
print(myhexstring)                                    # 68656c6c6f

print("\nDemonstrating the raw bytes mode of constructing a bit vector (useful for reading public and private keys):")
mypubkey = 'ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEA5amriY96HQS8Y/nKc8zu3zOylvpOn3vzMmWwrtyDy+aBvns4UC1RXoaD9rDKqNNMCBAQwWDsYwCAFsrBzbxRQONHePX8lRWgM87MseWGlu6WPzWGiJMclTAO9CTknplG9wlNzLQBj3dP1M895iLF6jvJ7GR+V3CRU6UUbMmRvgPcsfv6ec9RRPm/B8ftUuQICL0jt4tKdPG45PBJUylHs71FuE9FJNp01hrj1EMFObNTcsy9zuis0YPyzArTYSOUsGglleExAQYi7iLh17pAa+y6fZrGLsptgqryuftN9Q4NqPuTiFjlqRowCDU7sSxKDgU7bzhshyVx3+pzXO4D2Q== kak@pixie'
import base64
if sys.version_info[0] == 3:
    import binascii
    keydata = base64.b64decode(bytes(mypubkey.split(None)[1], 'utf-8'))
else:
    keydata = base64.b64decode(mypubkey.split(None)[1])
bv = BitVector( rawbytes = keydata )
print(bv)

# Test array-like indexing for a bit vector:
bv = BitVector( bitstring = '110001' )
print("\nPrints out bits individually from bitstring 110001:")
print(bv[0], bv[1], bv[2], bv[3], bv[4], bv[5])       # 1 1 0 0 0 1
print("\nSame as above but using negative array indexing:")
print(bv[-1], bv[-2], bv[-3], bv[-4], bv[-5], bv[-6]) # 1 0 0 0 1 1

# Test setting bit values with positive and negative
# accessors:
bv = BitVector( bitstring = '1111' )
print("\nBitstring for 1111:")
print(bv)                                    # 1111

print("\nReset individual bits of above vector:")
bv[0]=0;bv[1]=0;bv[2]=0;bv[3]=0
print(bv)                                    # 0000
print("\nDo the same as above with negative indices:")
bv[-1]=1;bv[-2]=1;bv[-4]=1
print(bv)                                    # 1011

print("\nCheck equality and inequality ops:")
bv1 = BitVector( bitstring = '00110011' )
bv2 = BitVector( bitlist = [0,0,1,1,0,0,1,1] )
print(bv1 == bv2)                           # True
print(bv1 != bv2)                           # False
print(bv1 < bv2)                            # False
print(bv1 <= bv2)                           # True
bv3 = BitVector( intVal = 5678 )
print(bv3.intValue())                       # 5678
print(bv3)                                  # 10110000101110
print(bv1 == bv3)                           # False
print(bv3 > bv1)                            # True
print(bv3 >= bv1)                           # True

# Write a bit vector to a file like object
fp_write = io.StringIO()
bv.write_bits_to_fileobject( fp_write )
print("\nGet bit vector written out to a file-like object:")
print(fp_write.getvalue())                  # 1011

print("\nExperiments with bitwise logical operations:")
bv3 = bv1 | bv2
print(bv3)                                  # 00110011
bv3 = bv1 & bv2
print(bv3)                                  # 00110011
bv3 = bv1 + bv2
print(bv3)                                  # 0011001100110011
bv4 = BitVector( size = 3 )
print(bv4)                                  # 000
bv5 = bv3 + bv4
print(bv5)                                  # 0011001100110011000
bv6 = ~bv5
print(bv6)                                  # 1100110011001100111
bv7 = bv5 & bv6
print(bv7)                                  # 0000000000000000000
bv7 = bv5 | bv6
print(bv7)                                  # 1111111111111111111

print("\nTry logical operations on bit vectors of different sizes:")
print(BitVector( intVal = 6 ) ^ BitVector( intVal = 13 ))   # 1011
print(BitVector( intVal = 6 ) & BitVector( intVal = 13 ))   # 0100
print(BitVector( intVal = 6 ) | BitVector( intVal = 13 ))   # 1111

print(BitVector( intVal = 1 ) ^ BitVector( intVal = 13 ))   # 1100
print(BitVector( intVal = 1 ) & BitVector( intVal = 13 ))   # 0001
print(BitVector( intVal = 1 ) | BitVector( intVal = 13 ))   # 1101

print("\nExperiments with setbit() and len():")
bv7[7] = 0
print(bv7)                                   # 1111111011111111111
print(len( bv7 ))                            # 19
bv8 = (bv5 & bv6) ^ bv7
print(bv8)                                   # 1111111011111111111

print("\nConstruct a bit vector from what is in the file testinput1.txt:")
bv = BitVector( filename = 'testinput1.txt' )
#print bv                                    # nothing to show
bv1 = bv.read_bits_from_file(64)
print("\nPrint out the first 64 bits read from the file:")
print(bv1)
     # 0100000100100000011010000111010101101110011001110111001001111001
print("\nRead the next 64 bits from the same file:")
bv2 = bv.read_bits_from_file(64)
print(bv2)
     # 0010000001100010011100100110111101110111011011100010000001100110
print("\nTake xor of the previous two bit vectors:")
bv3 = bv1 ^ (bv2)
print(bv3)
     # 0110000101000010000110100001101000011001000010010101001000011111

print("\nExperiment with dividing an even-sized vector into two:")
[bv4, bv5] = bv3.divide_into_two()
print(bv4)                            # 01100001010000100001101000011010
print(bv5)                            # 00011001000010010101001000011111

# Permute a bit vector:
print("\nWe will use this bit vector for experiments with permute()")
bv1 = BitVector( bitlist = [1, 0, 0, 1, 1, 0, 1] )
print(bv1)                                    # 1001101

bv2 = bv1.permute( [6, 2, 0, 1] )
print("\nPermuted and contracted form of the previous bit vector:")
print(bv2)                                    # 1010

print("\nExperiment with writing an internally generated bit vector out to a disk file:")
bv1 = BitVector( bitstring = '00001010' )
FILEOUT = open( 'test.txt', 'wb' )
bv1.write_to_file( FILEOUT )
FILEOUT.close()
bv2 = BitVector( filename = 'test.txt' )
bv3 = bv2.read_bits_from_file( 32 )
print("\nDisplay bit vectors written out to file and read back from the file and their respective lengths:")
print( str(bv1) + " " + str(bv3))
print(str(len(bv1)) + " " + str(len(bv3)))

print("\nExperiments with reading a file from the beginning to end:")
bv = BitVector( filename = 'testinput4.txt' )
print("\nHere are all the bits read from the file:")
while (bv.more_to_read):
    bv_read = bv.read_bits_from_file( 64 )
    print(bv_read)
print("\n")

print("\nExperiment with closing a file object and start extracting bit vectors from the file from the beginning again:")
bv.close_file_object()
bv = BitVector( filename = 'testinput4.txt' )
bv1 = bv.read_bits_from_file(64)
print("\nHere are all the first 64 bits read from the file again after the file object was closed and opened again:")
print(bv1)
FILEOUT = open( 'testinput5.txt', 'wb' )
bv1.write_to_file( FILEOUT )
FILEOUT.close()

print("\nExperiment in 64-bit permutation and unpermutation of the previous 64-bit bitvector:")
print("The permutation array was generated separately by the Fisher-Yates shuffle algorithm:")
bv2 = bv1.permute( [22, 47, 33, 36, 18, 6, 32, 29, 54, 62, 4,
                    9, 42, 39, 45, 59, 8, 50, 35, 20, 25, 49,
                    15, 61, 55, 60, 0, 14, 38, 40, 23, 17, 41,
                    10, 57, 12, 30, 3, 52, 11, 26, 43, 21, 13,
                    58, 37, 48, 28, 1, 63, 2, 31, 53, 56, 44, 24,
                    51, 19, 7, 5, 34, 27, 16, 46] )
print("Permuted bit vector:")
print(bv2)

bv3 = bv2.unpermute( [22, 47, 33, 36, 18, 6, 32, 29, 54, 62, 4,
                      9, 42, 39, 45, 59, 8, 50, 35, 20, 25, 49,
                      15, 61, 55, 60, 0, 14, 38, 40, 23, 17, 41,
                      10, 57, 12, 30, 3, 52, 11, 26, 43, 21, 13,
                      58, 37, 48, 28, 1, 63, 2, 31, 53, 56, 44, 24,
                      51, 19, 7, 5, 34, 27, 16, 46] )
print("Unpurmute the bit vector:")
print(bv3)

print("\nTry circular shifts to the left and to the right for the following bit vector:")
print(bv3)   # 0100000100100000011010000111010101101110011001110111001001111001
print("\nCircular shift to the left by 7 positions:")
bv3 << 7
print(bv3)   # 1001000000110100001110101011011100110011101110010011110010100000

print("\nCircular shift to the right by 7 positions:")
bv3 >> 7
print(bv3)   # 0100000100100000011010000111010101101110011001110111001001111001

print("Test len() on the above bit vector:")
print(len( bv3 ))                      # 64

print("\nTest forming a [5:22] slice of the above bit vector:")
bv4 = bv3[5:22]
print(bv4)                             # 00100100000011010

print("\nTest the iterator:")
for bit in bv4:
    print(bit)                         # 0 0 1 0 0 1 0 0 0 0 0 0 1 1 0 1 0

print("\nDemonstrate padding a bit vector from left:")
bv = BitVector( bitstring = '101010' )
bv.pad_from_left( 4 )
print(bv)                              # 0000101010

print("\nDemonstrate padding a bit vector from right:")
bv.pad_from_right( 4 )
print(bv)                              # 00001010100000

print("\nTest the syntax 'if bit_vector_1 in bit_vector_2' syntax:")
try:
    bv1 = BitVector( bitstring = '0011001100' )
    bv2 = BitVector( bitstring = '110011' )
    if bv2 in bv1:
        print("%s is in %s" % (bv2, bv1))
    else:
        print("%s is not in %s" % (bv2, bv1))
except ValueError as arg:
    print("Error Message: " + str(arg))

print("\nTest the size modifier when a bit vector is initialized with the intVal method:")
bv = BitVector( intVal = 45, size = 16 )
print(bv)                             # 0000000000101101
bv = BitVector( intVal = 0, size = 8 )
print(bv)                             # 00000000
bv = BitVector( intVal = 1, size = 8 )
print(bv)                             # 00000001

print("\nTesting slice assignment:")
bv1 = BitVector( size = 25 )
print("bv1= " + str(bv1))             # 0000000000000000000000000
bv2 = BitVector( bitstring = '1010001' )
print("bv2= " + str(bv2))             # 1010001
bv1[6:9]  = bv2[0:3]
print("bv1= " + str(bv1))             # 0000001010000000000000000
bv1[:5] = bv1[5:10]
print("bv1= " + str(bv1))             # 0101001010000000000000000
bv1[20:] = bv1[5:10]
print("bv1= " + str(bv1))             # 0101001010000000000001010
bv1[:] = bv1[:]
print("bv1= " + str(bv1))             # 0101001010000000000001010
bv3 = bv1[:]
print("bv3= " + str(bv3))             # 0101001010000000000001010

print("\nTesting reset function:")
bv1.reset( 1 )
print("bv1= " + str(bv1))             # 1111111111111111111111111
print(bv1[3:9].reset(0))              # 000000
print(bv1[:].reset(0))                # 0000000000000000000000000

print("\nTesting count_bit():")
bv = BitVector( intVal = 45, size = 16 )
y = bv.count_bits()
print(y)                              # 4
bv = BitVector( bitstring = '100111' )
print(bv.count_bits())                # 4
bv = BitVector( bitstring = '00111000' )
print(bv.count_bits())                # 3
bv = BitVector( bitstring = '001' )
print(bv.count_bits())                # 1
bv = BitVector( bitstring = '00000000000000' )
print(bv.count_bits())                # 0

print("\nTest setValue idea:")
bv = BitVector( intVal = 7, size =16 )
print(bv)                             # 0000000000000111
bv.setValue( intVal = 45 )
print(bv)                             # 101101

print("\nTesting count_bits_sparse():")
bv = BitVector( size = 2000000 )
bv[345234] = 1
bv[233]=1
bv[243]=1
bv[18]=1
bv[785] =1
print("The number of bits set: " + str(bv.count_bits_sparse()))    # 5

print("\nTesting Jaccard similarity and distance and Hamming distance:")
bv1 = BitVector( bitstring = '11111111' )
bv2 = BitVector( bitstring = '00101011' )
print("Jaccard similarity: " + str(bv1.jaccard_similarity( bv2 ))) # 0.5
print("Jaccard distance: " + str(bv1.jaccard_distance( bv2 )))     # 0.5
print("Hamming distance: " + str(bv1.hamming_distance( bv2 )))     # 4

print("\nTesting next_set_bit():")
bv = BitVector( bitstring = '00000000000001' )
print(bv.next_set_bit(5))                                    # 13
bv = BitVector( bitstring = '000000000000001' )
print(bv.next_set_bit(5))                                    # 14
bv = BitVector( bitstring = '0000000000000001' )
print(bv.next_set_bit(5))                                    # 15
bv = BitVector( bitstring = '00000000000000001' )
print(bv.next_set_bit(5))                                    # 16

print("\nTesting rank_of_bit_set_at_index():")
bv = BitVector( bitstring = '01010101011100' )
print(bv.rank_of_bit_set_at_index( 10 ))                     # 6

print("\nTesting isPowerOf2():")
bv = BitVector( bitstring = '10000000001110' )
print("int value: " + str(int(bv)))                          # 826
print(bv.isPowerOf2())                                       # False
print("\nTesting isPowerOf2_sparse():")
print(bv.isPowerOf2_sparse())                                # False

print("\nTesting reverse():")
bv = BitVector( bitstring = '0001100000000000001' )
print("original bv: " + str(bv))             # 0001100000000000001
print("reversed bv: " + str(bv.reverse()))   # 1000000000000011000

print("\nTesting Greatest Common Divisor (gcd):")
bv1 = BitVector( bitstring = '01100110' )
print("first arg bv: " + str(bv1) + " of int value: " + str(int(bv1))) #102
bv2 = BitVector( bitstring = '011010' )
print("second arg bv: " + str(bv2) + " of int value: " + str(int(bv2)))# 26
bv = bv1.gcd( bv2 )
print("gcd bitvec is: " + str(bv) + " of int value: " + str(int(bv)))  # 2

print("\nTesting multiplicative_inverse:")
bv_modulus = BitVector( intVal = 32 )
print("modulus is bitvec: " + str(bv_modulus) + " of int value: " + str(int(bv_modulus)))
bv = BitVector( intVal = 17 )
print("bv: " + str(bv) + " of int value: " + str(int(bv)))
result = bv.multiplicative_inverse( bv_modulus )
if result is not None:
    print("MI bitvec is: " + str(result) + " of int value: " + str(int(result)))
else: print("No multiplicative inverse in this case")
                                                  # 17

print("\nTest multiplication in GF(2):")
a = BitVector( bitstring='0110001' )
b = BitVector( bitstring='0110' )
c = a.gf_multiply(b)
print("Product of a=" + str(a) + " b=" + str(b) + " is " + str(c))
                                                  # 10100110

print("\nTest division in GF(2^n):")
mod = BitVector( bitstring='100011011' )          # AES modulus
n = 8
a = BitVector( bitstring='11100010110001' )
quotient, remainder = a.gf_divide(mod, n)
print("Dividing a=" + str(a) + " by mod=" + str(mod) + " in GF(2^8) returns the quotient " + str(quotient) + " and the remainder " + str(remainder))

print("\nTest modular multiplication in GF(2^n):")
modulus = BitVector( bitstring='100011011' )     # AES modulus
n = 8
a = BitVector( bitstring='0110001' )
b = BitVector( bitstring='0110' )
c = a.gf_multiply_modular(b, modulus, n)
print("Modular product of a=" + str(a) + " b=" + str(b) + " in GF(2^8) is " + str(c))

print("\nTest multiplicative inverses in GF(2^3) with " + \
                               "modulus polynomial = x^3 + x + 1:")
print("Find multiplicative inverse of a single bit array")
modulus = BitVector( bitstring='100011011' )     # AES modulus
n = 8
a = BitVector( bitstring='00110011' )
mi = a.gf_MI(modulus,n)
print("Multiplicative inverse of " + str(a) + " in GF(2^8) is " + str(mi))

print("\nIn the following three rows shown, the first row shows the " +\
      "\nbinary code words, the second the multiplicative inverses," +\
      "\nand the third the product of a binary word with its" +\
      "\nmultiplicative inverse:\n")
mod = BitVector( bitstring = '1011' )
n = 3
bitarrays = [BitVector(intVal=x, size=n) for x in range(1,2**3)]
mi_list = [x.gf_MI(mod,n) for x in bitarrays]
mi_str_list = [str(x.gf_MI(mod,n)) for x in bitarrays]
print("bit arrays in GF(2^3): " + str([str(x) for x in bitarrays]))
print("multiplicati_inverses: " +  str(mi_str_list))

products = [ str(bitarrays[i].gf_multiply_modular(mi_list[i], mod, n)) \
                    for i in range(len(bitarrays)) ]
print("bit_array * multi_inv: " + str(products))

# UNCOMMENT THE FOLLOWING LINES FOR
# DISPLAYING ALL OF THE MULTIPLICATIVE
# INVERSES IN GF(2^8) WITH THE AES MODULUS:

#    print("\nMultiplicative inverses in GF(2^8) with "  + \
#                      "modulus polynomial x^8 + x^4 + x^3 + x + 1:")
#    print("\n(This may take a few seconds)\n")
#    mod = BitVector( bitstring = '100011011' )
#    n = 8
#    bitarrays = [BitVector(intVal=x, size=n) for x in range(1,2**8)]
#    mi_list = [x.gf_MI(mod,n) for x in bitarrays]
#    mi_str_list = [str(x.gf_MI(mod,n)) for x in bitarrays]
#    print("\nMultiplicative Inverses:\n\n" + str(mi_str_list))
#    products = [ str(bitarrays[i].gf_multiply_modular(mi_list[i], mod, n)) \
#                        for i in range(len(bitarrays)) ]
#    print("\nShown below is the product of each binary code word " +\
#                     "in GF(2^3) and its multiplicative inverse:\n\n")
#    print(products)

print("\nExperimenting with runs():")
bv = BitVector( bitlist = (1, 0, 0, 1) )
print("For bit vector: " + str(bv))
print("       the runs are: " + str(bv.runs()))
bv = BitVector( bitlist = (1, 0) )
print("For bit vector: " + str(bv))
print("       the runs are: " + str(bv.runs()))
bv = BitVector( bitlist = (0, 1) )
print("For bit vector: " + str(bv))
print("       the runs are: " + str(bv.runs()))
bv = BitVector( bitlist = (0, 0, 0, 1) )
print("For bit vector: " + str(bv))
print("       the runs are: " + str(bv.runs()))
bv = BitVector( bitlist = (0, 1, 1, 0) )
print("For bit vector: " + str(bv))
print("       the runs are: " + str(bv.runs()))

print("\nExperiments with chained invocations of circular shifts:")
bv = BitVector( bitlist = (1,1, 1, 0, 0, 1) )
print(bv)
bv >> 1
print(bv)
bv >> 1 >> 1
print(bv)
bv = BitVector( bitlist = (1,1, 1, 0, 0, 1) )
print(bv)
bv << 1
print(bv)
bv << 1 << 1
print(bv)

print("\nExperiments with chained invocations of NON-circular shifts:")
bv = BitVector( bitlist = (1,1, 1, 0, 0, 1) )
print(bv)
bv.shift_right(1)
print(bv)
bv.shift_right(1).shift_right(1)
print(bv)
bv = BitVector( bitlist = (1,1, 1, 0, 0, 1) )
print(bv)
bv.shift_left(1)
print(bv)
bv.shift_left(1).shift_left(1)
print(bv)

# UNCOMMENT THE FOLLOWING LINES TO TEST THE
# PRIMALITY TESTING METHOD. IT SHOULD SHOW
# THAT ALL OF THE FOLLOWING NUMBERS ARE PRIME:
#    primes = [179, 233, 283, 353, 419, 467, 547, 607, 661, 739, 811, 877, \
#              947, 1019, 1087, 1153, 1229, 1297, 1381, 1453, 1523, 1597, \
#              1663, 1741, 1823, 1901, 7001, 7109, 7211, 7307, 7417, 7507, \
#              7573, 7649, 7727, 7841]
#    for p in primes:
#        bv = BitVector( intVal = p )
#        check = bv.test_for_primality()
#        print("The primality test for " + str(p) + ": " + str(check))

print("\nGenerate 32-bit wide candidate for primality testing:")
bv = BitVector( intVal = 0 )
bv = bv.gen_rand_bits_for_prime(32)
print(bv)
check = bv.test_for_primality()
print("The primality test for " + str(int(bv)) + ": " + str(check))

